pkg/controller/networkpolicy/clusternetworkpolicy.go

```diff
@@ -484,21 +561,136 @@ func serviceAccountNameToPodSelector(saName string) *metav1.LabelSelector {
 func hasPerNamespaceRule(cnp *crdv1beta1.ClusterNetworkPolicy) bool {
 	for _, ingress := range cnp.Spec.Ingress {
 		for _, peer := range ingress.From {
-			if peer.Namespaces != nil && peer.Namespaces.Match == crdv1beta1.NamespaceMatchSelf {
+			if peer.Namespaces != nil {
 				return true
 			}
 		}
 	}
 	for _, egress := range cnp.Spec.Egress {
 		for _, peer := range egress.To {
-			if peer.Namespaces != nil && peer.Namespaces.Match == crdv1beta1.NamespaceMatchSelf {
+			if peer.Namespaces != nil {
 				return true
 			}
 		}
 	}
 	return false
 }
 
+func namespaceRuleLabelKeys(cnp *crdv1beta1.ClusterNetworkPolicy) sets.Set[string] {
+	keys := sets.New[string]()
+	for _, ingress := range cnp.Spec.Ingress {
+		for _, peer := range ingress.From {
+			if peer.Namespaces != nil {
+				for _, k := range peer.Namespaces.SameLabels {
+					keys.Insert(k)
+				}
+			}
+		}
+	}
+	for _, egress := range cnp.Spec.Egress {
+		for _, peer := range egress.To {
+			if peer.Namespaces != nil {
+				for _, k := range peer.Namespaces.SameLabels {
+					keys.Insert(k)
+				}
+			}
+		}
+	}
+	return keys
+}
+
+func (n *NetworkPolicyController) getNamespaceLabels(ns string) map[string]string {
+	namespace, _ := n.namespaceLister.Get(ns)
+	return namespace.Labels
+}
+
+// groupNamespaceByLabelValue groups Namespaces if they have the same label value for all the
+// label keys listed. If a Namespace is missing at least one of the label keys, it will be
+// not be grouped. Example:
+//
+//	  ns1: app=web, tier=test, tenant=t1
+//	  ns2: app=web, tier=test, tenant=t2
+//	  ns3: app=web, tier=production, tenant=t1
+//	  ns4: app=web, tier=production, tenant=t2
+//	  ns5: app=db, tenant=t1
+//	  labelKeys = [app, tier]
+//
+//	Result after grouping:
+//	  "web,test,":       [ns1, ns2]
+//	  "web,production,": [ns3, ns4]
+func groupNamespacesByLabelValue(affectedNSAndLabels map[string]map[string]string, labelKeys []string) map[string][]string {
+	nsGroupedByLabelVal := map[string][]string{}
+	for ns, nsLabels := range affectedNSAndLabels {
+		if groupKey := getLabelValues(nsLabels, labelKeys); groupKey != labelValueUndefined {
```

Comment:
When the Namespace has no labels, this returned groupKey will be empty which is not equal to `labelValueUndefined`, all NSs with empty label will be added here.

Reply:
Good point, however since 1.22 all Namespaces will have default label added automatically. I believe for earlier versions Antrea will also add label based on Namespace name for selection purposes. https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#automatic-labelling